File: rw.cpp
#include <TMB.hpp>
template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_ARRAY(obs);    /* timeSteps x stateDim */
  PARAMETER_ARRAY(X); /* State */
  PARAMETER(transf_rho);
  PARAMETER_VECTOR(logsds);
  PARAMETER_VECTOR(logsdObs);
  int timeSteps = obs.dim[1];
  int stateDim  = obs.dim[0];
  // Parameter transform
  Type rho = 2.0 / (1.0 + exp(-2.0 * transf_rho)) - 1.0;
  vector<Type> sds   = exp(logsds);
  vector<Type> sdObs = exp(logsdObs);
  // Setup the covariance matrix Sigma
  matrix<Type> cov(stateDim, stateDim);
  for(int i = 0; i < stateDim; i++)
    for(int j = 0; j < stateDim; j++)
      cov(i,j) = pow(rho, Type(abs(i-j))) * sds[i] * sds[j];
  // Setup the multivariate normal density
  using namespace density;
  MVNORM_t<Type> neg_log_density(cov);
  // Define likelihood
  Type ans = 0;
  // Process likelihood: Draw initial state
  ans -= sum(dnorm(vector<Type>(X.col(0)),
		   Type(0),
		   Type(1),
		   true));
  // Process likelihood: Draw states
  for(int i=1; i<timeSteps; i++)    
    ans += neg_log_density(X.col(i) - X.col(i-1)); 
  // Data likelihood: Draw the data given the states
  for(int i = 1; i < timeSteps; i++)
    ans -= sum(dnorm(vector<Type>(obs.col(i)),
		     vector<Type>(X.col(i)),
		     sdObs,
		     true)); 
  return ans;
}

File: rw.R
library(MASS)

simdata <- function(){
  set.seed(1);
  local({
    rho=0.9
    sds=seq(0.5,2,length=stateDim)
    sdObs=rep(1,stateDim);
    corrMat=matrix(0.0,stateDim,stateDim)
    for(i in 1:stateDim){
      for(j in 1:stateDim){
        corrMat[i,j] = rho^abs(i-j)
      }
    }
    Sigma=corrMat*(sds %o% sds)
    d=matrix(NA,timeSteps,stateDim)
    obs=d;
    ##init state
    d[1,] = rnorm(stateDim);
    i=1;
    obs[i,] = d[i,] + rnorm(stateDim,rep(0,stateDim),sdObs)
    for(i in 2:timeSteps){
      d[i,] = d[i-1,] + mvrnorm(1,rep(0,stateDim),Sigma=Sigma)
      obs[i,] = d[i,] + rnorm(stateDim,rep(0,stateDim),sdObs)
    }
    matplot(d,type="l",las=1,lty=1,xlab="Time",ylab="Simulated states and measurements")
    matpoints(obs,pch=1);
  },.GlobalEnv)
}
stateDim=3
timeSteps=100
simdata()
library(TMB)
compile("rw.cpp")
dyn.load("rw.so")
data <- list(obs=t(obs))
parameters <- list(
  X=data$obs*0,
  transf_rho=0.1,
  logsds=sds*0,
  logsdObs=sdObs*0
  )
obj <- MakeADFun(data,parameters,random="X",DLL="rw")
newtonOption(obj, smartsearch=FALSE)

obj$fn()
obj$gr()
system.time(opt <- do.call("optim",obj))
pl <- obj$env$parList() ## <-- List of predicted random effects
#matpoints(t(pl$X),type="l",col="blue",lwd=2)
#summary(as.mle(opt))
sdreport(obj)
